core/src/term/mod.rs

```diff
@@ -216,6 +216,21 @@ pub enum Term {
     #[serde(skip)]
     Type(Type),
 
+    /// A custom contract built using e.g. `std.contract.from_predicate`. Currently, custom
+    /// contracts can be partial identities (the most general form, which either blame or return
+    /// the value with potentital delayed checks burried inside) or a predicate. Ideally, both
+    /// would fall under then `CustomContract` node.
+    ///
+    /// For now, we only put predicates built using `std.contract.from_predicate` here.
```

Comment:
I'm a bit confused as to why we'd want to support putting both forms in this node. Why now just convert everything to the general form, so there's only one representation to worry about?

Reply:
You mean, generic custom contracts and predicates? Well, the most general form is also the one with the less information. Once you convert a predicate to a generic custom contract (exactly what `from_predicate` used to do before this PR), there's no coming back. So that won't help in implementing the `or` operator.

I think the whole idea of having this new node is to be more fine-grained, and be able to say not only "this is an entirely opaque and general function I know nothing about" but rather "this is a custom contract, supposed to be used in contract position, and furthermore it comes from predicate `xxx` or from the general partial identity form `yyy`". Converting everything to the most general form is exactly what we used to do (without a dedicated node), basically.

Reply:
Makes sense, thanks for the explanation!